lib/nmatrix/math.rb

```diff
@@ -232,6 +283,49 @@ def factorize_lu with_permutation_matrix=nil
     [t, FactorizeLUMethods.permutation_matrix_from(pivot)]
   end
 
+  #
+  # call-seq:
+  #     factorize_qr -> [Q,R]
+  #
+  # QR factorization of a matrix without column pivoting. Q is orthogonal  
+  # and R is upper triangular if input is square or upper trapezoidal if 
+  # input is rectangular.  
+  #
+  # Only works for dense matrices.
+  #
+  # * *Returns* :
+  #   - Array containing Q and R matrices
+  #
+  # * *Raises* :
+  #   - +StorageTypeError+ -> only implemented for desnse storage.
+  #   - +ShapeError+ -> Input must be a 2-dimensional matrix to have a QR decomposition.
+  #   - +ShapeError+ -> Input must have shape[0] > shape[1]. 
+  #
+  def factorize_qr
+    raise(NotImplementedError, "only implemented for dense storage") unless self.stype == :dense
+    raise(ShapeError, "Input must be a 2-dimensional matrix to have a QR decomposition") unless self.dim == 2
+    raise(ShapeError, "Input must have shape[0] > shape[1]") unless self.shape[0] >= self.shape[1]
```

Comment:
I noticed that mentioned as a constraint on most lecture notes / [Wikipedia](https://en.wikipedia.org/wiki/QR_decomposition#Rectangular_matrix). My understanding is that the result of geqrf on matrices with more columns than rows would not be a meaningful QR decomposition. I'll look at this after I'm done with the tests.

https://en.wikipedia.org/wiki/QR_decomposition#Rectangular_matrix